File: app/src/main/java/org/breezyweather/background/polling/work/worker/NormalUpdateWorker.java
package org.breezyweather.background.polling.work.worker;

import android.annotation.SuppressLint;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.hilt.work.HiltWorker;
import androidx.work.WorkerParameters;
import androidx.work.impl.utils.futures.SettableFuture;

import java.util.List;

import dagger.assisted.Assisted;
import dagger.assisted.AssistedInject;
import org.breezyweather.common.basic.models.Location;
import org.breezyweather.remoteviews.NotificationHelper;
import org.breezyweather.remoteviews.WidgetHelper;
import org.breezyweather.location.LocationHelper;
import org.breezyweather.weather.WeatherHelper;

@HiltWorker
public class NormalUpdateWorker extends AsyncUpdateWorker {

    @AssistedInject
    public NormalUpdateWorker(@Assisted @NonNull Context context,
                              @Assisted @NonNull WorkerParameters workerParams,
                              LocationHelper locationHelper,
                              WeatherHelper weatherHelper) {
        super(context, workerParams, locationHelper, weatherHelper);
    }

    @Override
    public void updateView(Context context, Location location) {
        WidgetHelper.updateWidgetIfNecessary(context, location);
    }

    @Override
    public void updateView(Context context, List<Location> locationList) {
        WidgetHelper.updateWidgetIfNecessary(context, locationList);
        NotificationHelper.updateNotificationIfNecessary(context, locationList);
    }

    @SuppressLint("RestrictedApi")
    @Override
    public void handleUpdateResult(SettableFuture<Result> future, boolean failed) {
        future.set(failed ? Result.retry() : Result.success());
    }
}

File: app/src/main/java/org/breezyweather/background/polling/work/worker/NormalUpdateWorker.kt
package org.breezyweather.background.polling.work.worker

import android.annotation.SuppressLint
import android.content.Context
import androidx.hilt.work.HiltWorker
import androidx.work.WorkerParameters
import androidx.work.impl.utils.futures.SettableFuture
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import org.breezyweather.common.basic.models.Location
import org.breezyweather.location.LocationHelper
import org.breezyweather.remoteviews.Notifications.updateNotificationIfNecessary
import org.breezyweather.remoteviews.Widgets.updateWidgetIfNecessary
import org.breezyweather.weather.WeatherHelper

@HiltWorker
class NormalUpdateWorker @AssistedInject constructor(
    @Assisted context: Context,
    @Assisted workerParams: WorkerParameters,
    locationHelper: LocationHelper,
    weatherHelper: WeatherHelper
) : AsyncUpdateWorker(context, workerParams, locationHelper, weatherHelper) {
    override fun updateView(context: Context, location: Location) {
        updateWidgetIfNecessary(context, location)
    }

    override fun updateView(context: Context, locationList: List<Location>) {
        updateWidgetIfNecessary(context, locationList)
        updateNotificationIfNecessary(context, locationList)
    }

    @SuppressLint("RestrictedApi")
    override fun handleUpdateResult(future: SettableFuture<Result>?, failed: Boolean) {
        future?.set(if (failed) Result.retry() else Result.success())
    }
}
